File: frontend/src/lib/types/intake.ts
// Types for the procurement Intake surface. Mirrors the JSON returned by the
// `/api/intake` endpoints (backend `IntakeRequestResponse`). Date/datetime
// fields are ISO strings.
//
// **Money is `MoneyAmount` on the way in, `MoneyString` on the way out.** The
// backend serialises `estimated_amount` as a JSON number (`float(...)`), so
// `MoneyAmount` is the honest read type — and it makes `a - b` / `Math.max()`
// on the figure a type error rather than a convention. Outbound it is the exact
// decimal string the user typed (`utils/moneyInput.ts`), because a fractional
// JSON number is already a float by the time pydantic sees it. The `total` on
// the list/summary envelopes stays `number`: it is a ROW COUNT, not money.
// See `frontend/CLAUDE.md` § Money formatting.

import type { BadgeTone } from '$lib/components/ui/Badge.svelte';
import type { MessageKey } from '$lib/i18n/messages';
import type { MoneyAmount, MoneyString } from '$lib/utils/money';

export type IntakeStatus =
	| 'open'
	| 'in_review'
	| 'approved'
	| 'rejected'
	| 'converted'
	| 'cancelled';

export const INTAKE_STATUSES: IntakeStatus[] = [
	'open',
	'in_review',
	'approved',
	'rejected',
	'converted',
	'cancelled'
];

/**
 * The i18n key carrying each status label — never the English string itself.
 *
 * Both surfaces that render an intake status (the `/intake` list page and
 * `IntakeModal`) are inside the i18n extraction slice, so a hardcoded English
 * map here put a translated Reopen confirm ("Zurück auf „Offen“") directly
 * beside an untranslated `Open` badge. Keyed the same way
 * `notification.ts::EVENT_LABEL_KEYS` is; `Record<IntakeStatus, MessageKey>`
 * makes a new status a compile error rather than a blank badge, and
 * `intake.test.ts` proves every key exists in the catalogue.
 */
export const INTAKE_STATUS_LABEL_KEYS: Record<IntakeStatus, MessageKey> = {
	open: 'intake.status.open',
	in_review: 'intake.status.inReview',
	approved: 'intake.status.approved',
	rejected: 'intake.status.rejected',
	converted: 'intake.status.converted',
	cancelled: 'intake.status.cancelled'
};

// Badge tone per status, so the list page and the modal can't tint the same
// status two different shades — which is exactly what they did (.12 alpha on
// the list, .15 in the modal, and `open` already token-based in only one).
export const INTAKE_STATUS_TONES: Record<IntakeStatus, BadgeTone> = {
	open: 'accent',
	in_review: 'warning',
	approved: 'success',
	rejected: 'danger',
	// Converted means the ask left intake and became a requisition — a
	// mid-pipeline handoff, which is exactly what the `erp` purple marks.
	converted: 'erp',
	// Cancelled is the absence of a signal, not a weak one — flat, not tinted.
	cancelled: 'neutral'
};

/**
 * `IntakeRequest.status` is typed `string` (the API is the source of truth, and
 * the backend may add a status before this union does), so both call sites read
 * the tone through a tolerant accessor — mirroring how they read the label.
 * An unrecognised status renders flat rather than borrowing another tone's
 * meaning.
 */
export function intakeStatusTone(status: string): BadgeTone {
	return INTAKE_STATUS_TONES[status as IntakeStatus] ?? 'neutral';
}

/**
 * The message key for a status, or `null` for one this frontend doesn't know
 * (the caller renders the raw value — visible and searchable — rather than a
 * blank badge). Mirrors `intakeStatusTone`'s tolerance: `IntakeRequest.status`
 * is typed `string` because the API is the source of truth.
 */
export function intakeStatusLabelKey(status: string): MessageKey | null {
	return INTAKE_STATUS_LABEL_KEYS[status as IntakeStatus] ?? null;
}

export type IntakeType = 'software' | 'services' | 'hardware' | 'other';

export const INTAKE_TYPES: IntakeType[] = ['software', 'services', 'hardware', 'other'];

// Same treatment as the status labels above: the type name is a filter-chip
// label, a table cell AND the `{type}` interpolated into the translated
// questionnaire heading, so an English literal here shows up mid-sentence in a
// German modal.
export const INTAKE_TYPE_LABEL_KEYS: Record<IntakeType, MessageKey> = {
	software: 'intake.type.software',
	services: 'intake.type.services',
	hardware: 'intake.type.hardware',
	other: 'intake.type.other'
};

/** The message key for a request type, or `null` for an unrecognised one. */
export function intakeTypeLabelKey(requestType: string): MessageKey | null {
	return INTAKE_TYPE_LABEL_KEYS[requestType as IntakeType] ?? null;
}

// Flexible questionnaire payload — free-form key/value answers, shape varies by
// request type. Persisted verbatim as JSONB on the backend.
export type IntakeFormData = Record<string, unknown>;

export interface IntakeRequest {
	id: string;
	request_number: string;
	title: string;
	request_type: string;
	requester_user_id: string;
	description: string | null;
	estimated_amount: MoneyAmount;
	currency: string;
	vendor_name: string | null;
	vendor_id: string | null;
	status: string;
	form_data: IntakeFormData | null;
	needed_by: string | null;
	justification: string | null;
	converted_requisition_id: string | null;
	converted_po_id: string | null;
	created_at: string;
	updated_at: string;
}

export interface IntakeListResponse {
	items: IntakeRequest[];
	/** Row count of the whole filtered set — NOT money. */
	total: number;
	page: number;
	page_size: number;
}

/**
 * Whole-set KPI rollup from `GET /api/intake/summary`, over the SAME
 * status/type/search filters the list ran with.
 *
 * `openCount` / `reviewCount` used to filter the LOADED page by status, so both
 * contradicted the whole-set `total` count beside them.
 */
export interface IntakeSummary {
	/** Row count of the whole filtered set — NOT money. */
	total: number;
	by_status: Record<string, number>;
}

// Request side. Money goes out as the EXACT DECIMAL STRING the user typed —
// never a JSON number, which `json.loads` would have already rounded to a
// float before any pydantic `Decimal` annotation could see it.
export interface IntakeCreate {
	request_number?: string | null;
	title: string;
	request_type: string;
	description: string | null;
	estimated_amount: MoneyString | null;
	currency: string;
	vendor_name: string | null;
	form_data: IntakeFormData | null;
	needed_by: string | null;
	justification: string | null;
}

export interface IntakeUpdate {
	title?: string;
	request_type?: string;
	description?: string | null;
	estimated_amount?: MoneyString | null;
	currency?: string;
	vendor_name?: string | null;
	form_data?: IntakeFormData | null;
	needed_by?: string | null;
	justification?: string | null;
}

export interface IntakeConvertResponse {
	intake: IntakeRequest;
	requisition_id: string;
	requisition_number: string;
	created: boolean;
}

// Per-request-type questionnaire fields shown in the create modal. The labels
// drive a small dynamic form; answers are collected into `form_data`.
export const INTAKE_FORM_FIELDS: Record<IntakeType, { key: string; labelKey: MessageKey }[]> = {
	software: [
		{ key: 'seats', labelKey: 'intake.form.software.seats' },
		{ key: 'renewal', labelKey: 'intake.form.software.renewal' },
		{ key: 'data_residency', labelKey: 'intake.form.software.dataResidency' }
	],
	services: [
		{ key: 'scope', labelKey: 'intake.form.services.scope' },
		{ key: 'duration', labelKey: 'intake.form.services.duration' },
		{ key: 'sow_ref', labelKey: 'intake.form.services.sowRef' }
	],
	hardware: [
		{ key: 'quantity', labelKey: 'intake.form.hardware.quantity' },
		{ key: 'model', labelKey: 'intake.form.hardware.model' },
		{ key: 'ship_to', labelKey: 'intake.form.hardware.shipTo' }
	],
	other: [{ key: 'details', labelKey: 'intake.form.other.details' }]
};

File: backend/app/schemas/intake.py
"""Pydantic request/response schemas for the procurement intake router.

Intake captures a non-PO spend ask (software, services, hardware, other)
*before* a vendor/PO exists, routes it for review, and — once approved —
converts it into a ``PurchaseRequisition``.

Money convention (mirrors ``schemas/expense.py`` / ``schemas/contract.py``):
request fields are typed ``Decimal | None`` for exactness on the way in;
response/list fields serialise money as ``float | None`` (the router does
``float(...)``). Never ``float`` on a column or in-memory total. The flexible
questionnaire payload (``form_data``) is a free-form ``dict | None`` JSONB blob.
"""

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Field

from app.api.pagination import PageMeta
from app.models.procurement import IntakeStatus, IntakeType

# ---------------------------------------------------------------------------
# Intake requests
# ---------------------------------------------------------------------------


class IntakeRequestBase(BaseModel):
    title: str = Field(..., max_length=255)
    request_type: IntakeType = IntakeType.other
    description: str | None = None
    # Digits match `intake_requests.estimated_amount` Numeric(15, 2).
    estimated_amount: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    currency: str = Field(default="USD", max_length=3)
    vendor_name: str | None = Field(default=None, max_length=255)
    vendor_id: str | None = None
    # Flexible intake questionnaire answers (advisory; no PII).
    form_data: dict | None = None
    needed_by: date | None = None
    justification: str | None = None


class IntakeRequestCreate(IntakeRequestBase):
    # Optional caller-supplied number; the service generates one when omitted.
    request_number: str | None = Field(default=None, max_length=50)
    # Accepted for wire compatibility and then IGNORED — the requester is always
    # the authenticated caller. A converted requisition inherits this id and
    # approval checks segregation of duties against it, so it must not be the
    # creator's own input. Kept on the schema (rather than removed) so an old
    # client sending it gets an intake owned by itself, not a 422.
    requester_user_id: str | None = None


class IntakeRequestUpdate(BaseModel):
    """PATCH — every field optional. Only allowed while the intake is ``open``;
    ``status`` moves through the dedicated submit/approve/reject/cancel routes."""

    title: str | None = Field(default=None, max_length=255)
    request_type: IntakeType | None = None
    description: str | None = None
    # Digits match `intake_requests.estimated_amount` Numeric(15, 2).
    estimated_amount: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    currency: str | None = Field(default=None, max_length=3)
    vendor_name: str | None = Field(default=None, max_length=255)
    vendor_id: str | None = None
    form_data: dict | None = None
    needed_by: date | None = None
    justification: str | None = None


class IntakeRequestResponse(BaseModel):
    id: str
    request_number: str
    title: str
    request_type: str
    requester_user_id: str
    description: str | None
    estimated_amount: float | None
    currency: str
    vendor_name: str | None
    vendor_id: str | None
    status: str
    form_data: dict | None
    needed_by: str | None
    justification: str | None
    converted_requisition_id: str | None
    converted_po_id: str | None
    created_at: str
    updated_at: str


class IntakeRequestListResponse(PageMeta):
    items: list[IntakeRequestResponse]
    total: int


class IntakeDecision(BaseModel):
    """Optional body for an intake approve/reject/cancel — carries a reason.

    The reason is recorded in the audit row (and, on reject, stamped into the
    intake's ``form_data`` under ``review_reason`` so it survives on the row)."""

    reason: str | None = None


class IntakeConvertRequest(BaseModel):
    """Optional overrides applied when converting an approved intake into a
    requisition. When omitted, the intake's own title / estimated_amount /
    vendor / justification seed the single requisition line."""

    department: str | None = Field(default=None, max_length=120)
    needed_by: date | None = None


class IntakeConvertResponse(BaseModel):
    """The result of a convert-to-requisition call. ``created`` is ``False`` when
    the intake was already converted (idempotent replay) — the same
    ``requisition_id`` is returned either way."""

    intake: IntakeRequestResponse
    requisition_id: str
    requisition_number: str
    created: bool


# Re-exported enums so callers can import status/type from the schema module.
__all__ = [
    "IntakeStatus",
    "IntakeType",
    "IntakeRequestBase",
    "IntakeRequestCreate",
    "IntakeRequestUpdate",
    "IntakeRequestResponse",
    "IntakeRequestListResponse",
    "IntakeDecision",
    "IntakeConvertRequest",
    "IntakeConvertResponse",
]
